### PR TITLE
[AutoDiff] Fix SILGen JVP/VJP thunking bug.

### PR DESCRIPTION
**Pull request by dan-zheng**
SILGen JVP/VJP thunks for methods now always perform self reordering.
Add leak checking test.

Previously, self reordering was not performed if actual JVP/VJP type
matched expected JVP/VJP type, which is incorrect.

Gardening included:
- Change `SILFunction *SILGenFunction::getOrCreateAutoDiffLinearMapThunk` to
  `ManagedValue SILGenFunction::getThunkedAutoDiffLinearMap`. The new helper
  is more ergonomic and directly returns thunked values.
- Use ad-hoc `_vtable_entry_thunk` suffix for JVP/VJP vtable entry thunks.
- Add todo comments for [TF-685](https://bugs.swift.org/projects/TF/issues/TF-685?filter=allopenissues): principled AD thunk mangling.

Resolves [TF-698](https://bugs.swift.org/projects/TF/issues/TF-698).

---

Reproducer:

```swift
struct TF_698 : Differentiable & AdditiveArithmetic {
  var x: Float
  init(_ x: Float) {
    self.x = x
  }

  @differentiable(jvp: jvpMultiplied, vjp: vjpMultiplied)
  func multiplied(with other: Self) -> Self {
    return Self(x: x * other.x)
  }

  // Helper function to shorten JVP/VJP definitions.
  static func *(lhs: Self, rhs: Self) -> Self {
    return lhs.multiplied(with: rhs)
  }

  func jvpMultiplied(with other: Self) -> (Self, (Self, Self) -> Self) {
    return (multiplied(with: other), { dself, dother in
      self * dother + other * dself
    })
  }

  func vjpMultiplied(with other: Self) -> (Self, (Self) -> (Self, Self)) {
    return (multiplied(with: other), { v in
      (v * other, v * self)
    })
  }
}
let v = TF_698(1)
print(pullback(at: TF_698(3), TF_698(10)) { x, y in x.multiplied(with: y) }(v))
// Before:           (TF_698(3), TF_698(10))
// After (expected): (TF_698(10), TF_698(3))
```

Fixes derivatives of `Tensor.concatenated(with:)` and `Tensor.++`: https://github.com/tensorflow/swift-apis/pull/403

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILGen/SILGenFunction.h [hunk 1]
```diff
@@ -1780,14 +1780,15 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
...
+  /// differentials/pullbacks returned by user-defined JVP/VJP functions, and
+  /// apply it to the given differential/pullback.
   ///
   /// If `reorderSelf` is true, reorder self so that it appears as:
   /// - The last parameter, for differentials.
   /// - The last result, for pullbacks.
-  SILFunction *getOrCreateAutoDiffLinearMapThunk(
-      AutoDiffAssociatedFunctionKind assocFnKind,
+  ManagedValue getThunkedAutoDiffLinearMap(
+      ManagedValue linearMap, AutoDiffAssociatedFunctionKind assocFnKind,
```
Comment (dan-zheng) [inline lib/SILGen/SILGenFunction.h]: Use the recently added `AutoDiffLinearMapEnum`:
```suggestion
      ManagedValue linearMap, AutoDiffLinearMapEnum linearMapKind,
```


Will address in a follow-up to unblock progress.
This PR fixes latent bug in nightly toolchains.

File: lib/SILGen/SILGenPoly.cpp [hunk 2]
```diff
@@ -3746,8 +3761,8 @@ SILGenModule::getOrCreateAutoDiffAssociatedFunctionThunk(
 
   SmallVector<SILValue, 8> directResults;
   extractAllElements(apply, thunkSGF.B, directResults);
-  auto linearMap = directResults.back();
-  auto linearMapFnType = linearMap->getType().castTo<SILFunctionType>();
+  auto linearMap = ManagedValue::forBorrowedObjectRValue(directResults.back());
+  auto linearMapFnType = linearMap.getType().castTo<SILFunctionType>();
   auto targetLinearMapFnType = thunk->mapTypeIntoContext(
       origAssocFnType->getResults().back().getSILStorageType())
           .castTo<SILFunctionType>();
```
Comment (rxwei) [inline lib/SILGen/SILGenPoly.cpp]: I just noticed that you are doing manual memory management below in SILGen. We should use SILGen ManagedValue operations instead. Please file an issue to track this future work.

File: lib/SILGen/SILGenPoly.cpp [hunk 3]
```diff
@@ -3746,8 +3761,8 @@ SILGenModule::getOrCreateAutoDiffAssociatedFunctionThunk(
 
   SmallVector<SILValue, 8> directResults;
   extractAllElements(apply, thunkSGF.B, directResults);
-  auto linearMap = directResults.back();
-  auto linearMapFnType = linearMap->getType().castTo<SILFunctionType>();
+  auto linearMap = ManagedValue::forBorrowedObjectRValue(directResults.back());
```
Comment (rxwei) [inline lib/SILGen/SILGenPoly.cpp]: You aren’t using any SILGen memory management in this thunk. This should be ‘ManagedValue::forUnmanaged’ or something close (I forgot the exact name).

**Review by rxwei — COMMENTED** [review]
The leak checking tests you added are valid, but they are almost never effective for the thunks you added. They do not test whether you leaked any partial_apply’s. You can either convert all of these thunks to use ownership (which should be relatively easy), or manually test these cases before you merge. Making these thunks use ownership is a must in the short term.

File: lib/SILGen/SILGenPoly.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILGen/SILGenPoly.cpp]: I have a WIP patch after this one that enables ownership for SILGen linear map thunks.

**Comment by dan-zheng** [comment]
> The leak checking tests you added are valid, but they are almost never effective for the thunks you added. They do not test whether you leaked any partial_apply’s. You can either convert all of these thunks to use ownership (which should be relatively easy), or manually test these cases before you merge. Making these thunks use ownership is a must in the short term.

I have a WIP patch to enable ownership in SILGen thunks. Is it okay for this patch land first to resolve [TF-698](https://bugs.swift.org/projects/TF/issues/TF-698)?

**Comment by rxwei** [comment]
Yep, resolving correctness is more important. Could you please run a few manual memory tests through Instruments before merging?

**Comment by dan-zheng** [comment]
> Yep, resolving correctness is more important. Could you please run a few manual memory tests through Instruments before merging?

I'm AFK now but I can use Instruments to test leaks in a few hours. (I understand the need to test `partial_apply` leaks, like in https://github.com/apple/swift/pull/26384, and why the current tests are ineffective.)

File: lib/SILGen/SILGenPoly.cpp (hunk 3 above)
Comment (dan-zheng) [reply lib/SILGen/SILGenPoly.cpp]: Will address shortly in a follow-up.
(I plan to enable ownership for AD associated function thunks soon.)

**Comment by dan-zheng** [comment]
My patch enabling ownership for AD linear map thunks is ready for review.
I'll merge this PR first (fixing correctness but not leaks) to separate orthogonal changes.

**Pull request merged**